### DISCUSSION TITLE
When is it convenient to have two or more containers in a single pod?

### DISCUSSION THREAD

**Discussion by sergiorivas** [BretFisher/ama#198] [Q&A]
I am not sure what's a situation for having more than one container in a pod.

For example:

If you have an app with (frontend & backend)

You can argue: it's better to have them in separate pods since they can be scaled individually, but for another side, when you create a new feature usually they have to be upgraded **at the same time**.

What do you recommend?

**Comment by BretFisher** [comment]
The reasons would be when two parts of the same app need to always be on the same host together.

Some valid uses:
1. nginx container using a socket file or localhost TCP to talk to php-fpm container
2. a container that writes files to local disk for processing, and a container that sees that same volume and processes those files and then puts them somewhere else. Say, a user image upload utility that shrinks user uploads in a blog before putting them in persistant storage
3. a "sidecar" container, like a logging tool that picks up the logs of another container sharing a volume and sends them to a datastore

More examples of sidecars:
* https://kubernetes.io/docs/concepts/workloads/pods/pod-overview/#understanding-pods
* https://kubernetes.io/docs/tasks/access-application-cluster/communicate-containers-same-pod-shared-volume/
* https://www.weave.works/blog/container-design-patterns-for-kubernetes/
* (other paterns besides sidecar) https://matthewpalmer.net/kubernetes-app-developer/articles/multi-container-pod-design-patterns.html
